File: npm/src/cjson.ts
import { isAbsolutePath, read } from "./utils/file";
import * as path from 'path';
import { Is } from "./utils/is";
import Keywords from "./utils/keywords";
import { Json, isContentJson } from "./utils/json";
import { AbsolutePathConstraintError, CJSONContentInsteadOfFilePath } from "./utils/errors";
import { refineRelativePaths, refineRuntimeVals } from "./utils/refinery";

/**
 * Coded JSON is an extended format of JSON formatted data storage, which gives
 * you more previledge to organize data into more structured format.
 * 
 * Here is an example for `CJSON` format:
 * @example
 * 
 * {
    "source": $import "./source.json",
    "target": {
        "fruit": "Apple",
        "size": "Large",
        "color": "Red",
        "secColor": $.target.color,
        "colorList": [ $.target.color, $.target.secColor ],
        // You can add comments like this
        "digitCheck": 1.5,
        "digitImport": $.target.digitCheck,
        "digitArrayImport": [ $.target.digitCheck, $.target.digitImport ]
    }
}
 * 
 * 
 * The above `CJSON` snipped will be deserialized in JSON format and can be used 
 * as same as other JSON files.
 * 
 * For other details, please refer to official page: {@link https://subhendushekhar.github.io/cjson/}
 */
export class Cjson extends Is {
    // private obj: any;
    private filePath: string ;
    private content: string = "";
    public json: Json | undefined = undefined;
    public isContentCJson: boolean | undefined;
    public isContentJson = (isFilePath: boolean): boolean => { return isContentJson(this.content, isFilePath) };

    /**
     * Coded JSON is an extended format of JSON formatted data storage, which gives
     * you more previledge to organize data into more structured format.
     * 
     * Here is an example for `CJSON` format:
     * @example
     * 
     * {
        "source": $import "./source.json",
        "target": {
            "fruit": "Apple",
            "size": "Large",
            "color": "Red",
            "secColor": $.target.color,
            "colorList": [ $.target.color, $.target.secColor ],
            // You can add comments like this
            "digitCheck": 1.5,
            "digitImport": $.target.digitCheck,
            "digitArrayImport": [ $.target.digitCheck, $.target.digitImport ]
        }
    }
     * 
     * 
     * The above `CJSON` snipped will be deserialized in JSON format and can be used 
     * as same as other JSON files.
     * 
     * For other details, please refer to official page: {@link https://subhendushekhar.github.io/cjson/}
     * 
     * @param content Can be path to the CJSON file. In this case the second param can be optional
     * @param isContentCJson Set this true if you are passing raw CJSON content as `content`
     */
    constructor(content: string, isContentCJson?: boolean) {
        super();
        this.obj = undefined;
        
        if(isAbsolutePath(content) && isContentCJson) throw CJSONContentInsteadOfFilePath("CJSON flag is true, but got file path");

        if(isContentCJson) {
            this.filePath = __dirname;
            this.content = content;
            this.isContentCJson = isContentCJson;
        } else {
            this.filePath = content;
            this.content = read(this.filePath);
        }
        this.decodeKeywords();
    }
    /**
     * Root function for decoding keywords
     * Need to improve performance. `v1.0.0`
     */
    private decodeKeywords() {
        var isChanged: boolean = false;
        while(true) {
            isChanged = false;
            if(this.isImport(this.content)) {
                this.content = this.decodeImport(this.content, path.dirname(this.filePath));
                isChanged = true
            }
            if(this.isSingleLineComment(this.content)) {
                this.decodeSingleLineComment(this.content);
                isChanged = true
            }
            this.content = this.decodeRuntimeKeys(this.content);
            if(!isChanged) break;
        }
        if(this.isRelativeJPath(this.content)) {
            this.content = this.decodeRelativePaths(this.content);
            isChanged = true;
        }
    }
    private refineObj(content?: string) {
        if(content) this.content = content;

        this.json = new Json(this.content, false);
        this.obj = JSON.parse(this.content);
    }
    /**
     * Import functions path to relative file is deocded.
     * Modifies `content`
     */
    private decodeRelativePaths(content: string) {
        var uniqueKeys = content.match(Keywords.relativeJPathRegex)?.filter((value, index, array) => { return array.indexOf(value) === index } )

        if(uniqueKeys) 
            content = refineRelativePaths(content, uniqueKeys);

        this.refineObj(content);
        return content;
    }
    /**
     * Deserializes the keywords.
     * @returns `JSON` if no errors. Else `undefined`
     */
    public deserialize() : Cjson {
        this.content = this.decodeRelativePathValues(this.content);
        this.refineObj(this.content);
        return this.obj;
    }
    /**
     * Returns file path from `import` keyword
     * @param lineItem Comma separated line item in string
     * @returns File path in string
     */
    private getFilePath(lineItem: string): string {
        return lineItem.split(Keywords.importKey)[1].split("\"")[0];
    }
    /**
     * Decodes `import` keyword
     * @param lineItem Comma separated line item in string
     */
    private decodeImport(content: string, curPath: string): string {
        var filePath: string = this.getFilePath(content);
        var importFilePath: string;

        if(isAbsolutePath(filePath))
            importFilePath = filePath;

        else if(!isAbsolutePath(filePath) && this.isContentCJson) throw AbsolutePathConstraintError("Only absolute path is supported in import statements");

        else 
            var importFilePath: string = path.join(curPath, filePath);
            
        var innerContent: string = read(importFilePath);

        var qr = "";
        if(this.isImport(innerContent))
            qr = this.decodeImport(innerContent, path.dirname(importFilePath))
        else 
            qr = innerContent;
        
        content = content.replace(Keywords.importKey + filePath + "\"", qr);

        if(this.isImport(content)) 
            content = this.decodeImport(content, curPath);

        return content;
    }
    /**
     * Identifies comment lines. Can identify multiple lined comments
     * @param lineItem Comma separated line item in string
     */
    private decodeSingleLineComment(lineItem: string) {
        let lineSplit: string[] = lineItem.split("\r\n");
        for(let i = 0; i < lineSplit.length; i ++) {
            if(lineSplit[i].trim() !== "" && lineSplit[i].trim().startsWith(Keywords.singleLineComment))
                this.content = this.content.replace(lineSplit[i], "");
        }
    }
    private decodeRuntimeVals(content: string, uniqueKeys: string[], injectingObj: any) {
        if(uniqueKeys !== undefined) {
            content = refineRuntimeVals(content, uniqueKeys);
            
            uniqueKeys?.map(eachKey => {
                eachKey = eachKey.split("<")[1].split(">")[0];

                let keyRegex = new RegExp("<-" + eachKey + "->", 'g');
                
                if(typeof injectingObj[eachKey] === "string")
                    content = content.replace(keyRegex, injectingObj[eachKey]);
                else {
                    keyRegex = new RegExp("\"<-" + eachKey + "->\"", 'g');
                    content = content.replace(keyRegex, JSON.stringify(injectingObj[eachKey]));
                }
            });
        }
        return content;
    }
    /**
     * Use this for injecting variable at runtime.
     * 
     * Value need to be replaced can be stored with a key like `<key>`.
     * 
     * Pass the value as json object in `injectingObj`
     * @param injectingObj Runtime values to be injected in json format
     * @returns `JSON` if no errors. Else `undefined`
     */
    public inject(injectingObj: any) {

        let content = this.content;

        var uniqueKeys = this.content.match(Keywords.decodedRuntimeKeys)?.filter((value, index, array) => { return array.indexOf(value) === index });
        
        uniqueKeys = uniqueKeys?.flatMap(eachElem => eachElem.split("<-")[1].split("->")[0]);

        var injectObjKeys: string[] = Object.keys(injectingObj);

        for(let i = 0; i < injectObjKeys.length; i ++) {
            if(uniqueKeys?.includes(injectObjKeys[i])) {
                if(typeof injectingObj[injectObjKeys[i]] === "boolean" || 
                    typeof injectingObj[injectObjKeys[i]] === "bigint" ||
                    typeof injectingObj[injectObjKeys[i]] === "number")
                    content = content.replace(new RegExp("\"<-" + injectObjKeys[i] + "->\"", "g"), injectingObj[injectObjKeys[i]]);
                else if(typeof injectingObj[injectObjKeys[i]] === "object")
                    content = content.replace(new RegExp("\"<-" + injectObjKeys[i] + "->\"", "g"), JSON.stringify(injectingObj[injectObjKeys[i]]));
                else
                    content = content.replace(new RegExp("<-" + injectObjKeys[i] + "->", "g"), injectingObj[injectObjKeys[i]]);
                
                if(this.decodedRuntimeKeyList !== undefined)
                    this.removeFromStringArray(this.decodedRuntimeKeyList, injectObjKeys[i]);
                else
                    console.error("Cannot update Base.decodedRuntimeKeyList as it is undefined");
            }
        }

        if(this.content.match(Keywords.decodedRuntimeKeys)?.filter((value, index, array) => { return array.indexOf(value) === index }).length !== 0)
            console.warn("Still some runtime keys are expecting values... Run cjson.getAllRuntimeKeys()")
        else 
            this.isInjectExist = false;

        content = this.decodeRelativePathValues(content);
        this.refineObj(content);
        return this;
    }
    /**
     * Converts JSON object to string. Just a wrapper over `JSON.stringify()`
     * @param obj JSON object
     * @returns JSON string
     */
    public static toString(obj: any) {
        if(obj === null) return "{}";
        else if(!isContentJson(JSON.stringify(obj))) throw new Error("Object is not a JSON");
        else return JSON.stringify(obj);
    }
    /**
     * Deserializes `CJSON` content and returns content as string.
     * 
     * Content will be of pure JSON content and can be parsed as `JSON`
     * @returns `JSON` equivalent of `CJSON` content in `string`
     */
    public deserializeAsString() : string {
        this.deserialize();
        return this.content;
    }
    /**
     * Removes a key:value from the CJSON context. Key will be JPath in `$.full.path` format
     * 
     * The function automatically deserializes before removing. So, no need to explicitely deserialize it.
     * @param key JPath to the key to be removed.
     * @returns Resultant content in `JSON` object
     */
    public remove(key: string) {
        this.obj = this.json?.removeWithKey(key, this.content);
        this.content = Cjson.toString(this.obj);
        return this;
    }
    /**
     * Replace a JPath to a specified value. The function context is of JSON and cannot be used in CJSON context.
     * 
     * In order to use this in CJSON context, follow below steps:
     * <ol>
     * <li>Create CJson object</li>
     * <li>Deserialize</li>
     * <li>Deserialize</li>
     * <li>cjson.json?.replace("$.jpath", value, object)</li>
     * </ol>
     * @param key 
     * @param value 
     * @param jsonObject 
     * @returns 
     */
    public replace = (jPath: string, value: any) => {
        if(jPath.startsWith("$."))
            jPath = jPath.split("$.")[1];
        this.obj = this.json?.replace(jPath, value, this.obj);
        this.refineObj(Cjson.toString(this.obj));
        return this;
    }
    private decodeRuntimeKeys(content: string) {
        var runtimeKeys: string[] | undefined = content.match(Keywords.runtimeVals)?.filter((value, index, array) => { return array.indexOf(value) === index });

        if(runtimeKeys !== undefined && runtimeKeys !== null) {
            this.isInjectExist = true;

            for(let i = 0; i < runtimeKeys.length; i ++) {
                let variable: string = runtimeKeys[i].split("<")[1].split(">")[0];
                let ignoreUnderQuotes: RegExpMatchArray | null = content.match("\".*" + runtimeKeys[i] + ".*\"");

                if(ignoreUnderQuotes === null) {
                    variable = "\"<-" + variable + "->\"";
                    content = content.replace(new RegExp(runtimeKeys[i], "g"), variable);
                }
            }
        }
        return content;
    }
    private decodeRelativePathValues(content: string) {
        var encodedKeys: string[] | undefined = content.match(Keywords.encodedRelativeJPathRegex)?.filter((value, index, array) => { return array.indexOf(value) === index });
        if(encodedKeys) {
            for(let i = 0; i < encodedKeys.length; i ++) {
                let key: string = encodedKeys[i].split("<")[1].split(">")[0];
                let value = this.json?.parse(key);
                if(value !== null) {
                    while(JSON.stringify(value).includes("$.")) 
                        value = this.json?.parse(value.split("<")[1].split(">")[0]);

                    if(typeof value === "string")
                        content = content.replace(new RegExp(encodedKeys[i].replace("$", "\\$"), "g"), value);
                    else
                        content = content.replace(new RegExp("\"" + encodedKeys[i].replace("$", "\\$") + "\"", "g"), value);
                }
                else
                    content = content.replace(new RegExp("\"" + encodedKeys[i].replace("$", "\\$") + "\"", "g"), "null");
            }
        }
        return content;
    }
}

File: python/cjson/src/cjson.py
from utils.file import read, is_path_absolute
from utils._is import Is
from utils.keywords import Keywords
from utils._json import Json, is_content_json as content_json_check
from os import path
import json, re
from utils._exceptions import AbsolutePathConstraintError, FilePathAndCJSONCotentConflict, UnexpectedCJSONContent

def is_content_json(content: str, is_file_path: bool = False):
        ''' Checks if the parsed content is JSON
        '''
        return content_json_check(content, is_file_path=is_file_path)

class Cjson(Is):
    __file_path: str
    __is_content_cjson: bool = False
    json: Json | None = None

    def __init__(self, content: str, is_content_cjson: bool = False):
        ''' Initializes and decodes `CJSON` files.

            This can also be used for parsing JSON files in `CJSON` way.

            `content` can be file path and cjson/json type string.

            Parsing in CJSON way unlocks many functions. For more details, see function documentation.

            Set `is_content_cjson` as `True` if `content` is raw CJSON content instead of file path
        '''
        super().__init__()
        self._obj = None

        if(is_path_absolute(content) and is_content_cjson):
            raise FilePathAndCJSONCotentConflict()
        elif is_content_cjson == True:
            self.__is_content_cjson = is_content_cjson
            self._content = content
            self.__file_path = None
        else:
            self.__file_path = content
            self._content = read(self.__file_path)
    
    def __decode_keywords(self):
        is_changed: bool = False
        while(True):
            is_changed = False

            if self._is_import(self._content):
                self._content = self.__decode_import(self._content)
                is_changed = True
            
            if self.is_single_line_comments(self._content):
                self.__decode_single_line_comment(self._content)
                is_changed = True
            
            if not is_changed:
                break
    
    def __refine_obj(self, content: str = None):
        if content != None:
            self._content = content
        
        self.json = Json(self._content, False)
        self._obj = json.loads(self._content)

    def __decode_relative_paths(self, content: str):
        path_keys: list[str] = re.findall(Keywords.relative_jpath_regex, content)
        unique_keys: list[str] = []

        for each_key in path_keys:
            if each_key not in unique_keys:
                unique_keys.append(each_key)

        for each_key in unique_keys:
            content = content.replace(each_key, "\"<" + each_key + ">\"")

        self.__refine_obj(content=content)
    
        for each_key in unique_keys:
            key_regex: str = "<" + each_key + ">"
            
            parsed_val: str = self.json.parse(each_key.split(Keywords.relative_jpath)[1])

            while(str(parsed_val).startswith("<" + Keywords.relative_jpath)):
                self.__refine_obj(content=content)
                parsed_val: str = self.json.parse(each_key.split(Keywords.relative_jpath)[1])
            
            if type(parsed_val) != str:
                content = content.replace("\"" + key_regex + "\"", str(parsed_val))
            else:
                content = content.replace(key_regex, parsed_val)

        self.__refine_obj(content=content)
    
    def deserialize(self):
        ''' Deserializes CJSON context and returns dictionary object in JSON context.

            Please Note, if any key is detected for runtime variable injections is
            replaced with `<-tag->`. No error/warning is thrown.

            To inject runtime variables, use `inject` function
        '''
        self.__decode_keywords()

        runtime_keys: list[str] = re.findall(Keywords.runtime_vals_regex, self._content)
        '''Call this object to unlock native JSON functions
        '''
        self._content = self.__refine_runtime_vals(self._content, runtime_keys=runtime_keys)
        self.json = Json(self._obj)
        
        self.__decode_relative_paths(self._content)

        ''' Returns the JSON compiled object for the given `CJSON` file. '''
        return self._obj
    
    def deserializeAsString(self):
        ''' Deserializes `CJSON` content and returns content as string.

            Content will be of pure JSON content and can be parsed as `JSON`
        '''
        if self._obj == None:
            self.deserialize()

        return self._content

    def __get_file_path(self, line_item: str):
        return line_item.split(Keywords.import_key)[1].split("\"")[0]

    def __decode_import(self, content: str, cur_path: str = None):
        global import_file_path
        file_path: str = self.__get_file_path(line_item=content)
        file_name: str = file_path.split("/")[len(file_path.split("/")) - 1]

        if(is_path_absolute(file_path=file_path)):
            import_file_path = file_path

        elif(not is_path_absolute(file_path=file_path) and self.__is_content_cjson):
            raise AbsolutePathConstraintError("Only absolute path is supported in import statements")
        else:
            dir_name: str = path.join(path.dirname(self.__file_path), path.dirname(file_path))

            if(cur_path != None):
                dir_name = path.join(cur_path, path.dirname(file_path))

            import_file_path = path.join(dir_name, file_name)
        
        content = content.replace(Keywords.import_key + file_path + "\"", read(import_file_path))

        if(self._is_import(content)):
            return self.__decode_import(content=content, cur_path=path.dirname(import_file_path))
        else:
            return content

    def __decode_single_line_comment(self, line_item: str):
        line_split: list[str] = line_item.split("\n")
        for i in range(0, len(line_split)):
            if line_split[i].strip() != "" and line_split[i].strip().startswith(Keywords.single_line_comment):
                self._content = self._content.replace(line_split[i], "")
    
    def __refine_runtime_vals(self, content: str, runtime_keys: list[str]):
        unique_keys: list[str] = []

        for each_runtime_keys in runtime_keys:
            if(each_runtime_keys not in unique_keys):
                unique_keys.append(each_runtime_keys)
        
        for each_runtime_keys in unique_keys:
            content = content.replace(each_runtime_keys, "\"<-" + each_runtime_keys.split("<")[1].split(">")[0] + "->\"")
        
        return content

    def __inject_with_content(self, content: str, key: str, value: any):
        if type(value) != str:
            content = content.replace("\"<-" + key + "->\"", json.dumps(value))
        else:
            content = content.replace("<-" + key + "->", str(value))
        return content

    def inject_with_key_value(self, key: str, value: str):
        content = self._content
        self.__decode_keywords()

        runtime_keys: list[str] = re.findall(Keywords.runtime_vals_regex, self._content)
        self._content = self.__refine_runtime_vals(content=self._content, runtime_keys=runtime_keys)

        '''Call this object to unlock native JSON functions
        '''
        self.json = Json(self._obj)
        self.__decode_relative_paths(self._content)

        content = self.__inject_with_content(content=content, key=key, value=value)

        return content

    def inject(self, injecting_obj: dict):
        self.__decode_keywords()

        runtime_keys: list[str] = re.findall(Keywords.runtime_vals_regex, self._content)
        
        self._content = self.__refine_runtime_vals(content=self._content, runtime_keys=runtime_keys)
        '''Call this object to unlock native JSON functions
        '''
        self.json = Json(self._obj)
        self.__decode_relative_paths(self._content)
        
        for each_key in injecting_obj.keys():
            self._content = self.__inject_with_content(self._content, key=each_key, value=injecting_obj[each_key])
        
        self.__refine_obj(self._content)

        return self._obj

    def __get_as_string(obj: any) -> str:
        if(obj == None):
            return "null"
        elif((type(obj) is int) or (type(obj) is float) or (type(obj) is bool)):
            return str(obj)
        elif(type(obj) is str):
            return "\"" + str(obj) + "\""
        elif(type(obj) is list):
            li: list = obj
            values: str = "["
            for objs in li:
                if(type(objs) is str):
                    values += "\"" + objs + "\","
                else:
                    values += Cjson.__get_as_string(obj=objs) + ","
            values = values[-1]
            return values + "]"
        elif(type(obj) is dict):
            values = "{"
            hashObj: dict = obj
            for key in hashObj.keys():
                if hashObj[key] != None:
                    if type(hashObj[key]) is str:
                        values += "\"" + key + "\":\"" + hashObj[key] + "\","
                    else:
                        values += "\"" + key + "\":" + Cjson.__get_as_string(hashObj[key]) + ","
                else:
                    return None
            return values[-1] + "}"
        else:
            values: str = "{"
            for key in obj.keys():
                if type(obj[key]) is str:
                    values += "\"" + key + "\":\"" + obj[key] + "\""
                else:
                    values += "\"" + key + "\":" + Cjson.__get_as_string(obj[key])
            values += "}"
            return values

    def to_string(obj: any) -> str:
        if(type(obj) is dict and type(obj) is list):
            raise UnexpectedCJSONContent()
        if obj == "":
            return "{}"
        return Cjson.__get_as_string(obj=obj)

    def remove(self, key: str):
        self.deserialize()
        self.json = self.json._remove_with_key(key, self._content)

        self._obj = self.json._obj
        self._content = self.json._content
        return self

    def replace(self, jpath: str, value: str):
        self.deserialize()
        if jpath.startswith("$."):
            jpath = jpath.replace("$.", "", 1)

        self._obj = self.json.replace(jpath=jpath, value=value, obj=self._obj)
        self._content = self.json._content
        return self
    
    def get_obj(self):
        return self._obj
    
    def get_string(self):
        return self._content
